### PR TITLE
Example of what automated PR doc checking could look like

### PR DESCRIPTION
**Pull request by JohnHillegass**
For #311

This was a joint effort with @JonZeolla. Looking forward to feedback and further discussion on how we can implement it.  

This PR adds a Makefile with document check commands which could be run locally or remotely on Github actions. It adds an editorconfig for editor level checks, linting (markdownlint) and link checking (markdown-link-check) for markdown files and a spellcheck (cspell).

To run these commands locally, you will need `docker-compose` installed and can then just run the commands listed in the Makefile. For example, to run the lint checks we would use `make lint` as a terminal command.

For implementation, I suggest setting branch protection rules as shown below so that they do not block PR merges but will call out any issues. The `setup` step just prints out what files were identified as changed and should always succeed. 

image

An example PR check using these via Github actions can be seen here
https://github.com/JohnHillegass/sig-security/pull/4

The checks will show red when an issue is detected
image

Upon digging into the errors we can see what exactly failed in the CI outputs. 

For example, here we see the identified spelling issues
image

Here is another interesting issue identified in the link checker. It appears this file should be referenced as `CODEOWNERS` instead of `CODEOWNERS.md` maybe the file name changed recently? Either way, it's a good example of actionable doc check info that we could use as PR reviewers. 
image

A nice perk of using Github Actions is that the collaborator will also be emailed with a summary of the CI checks so they will have a prompt to go ahead and fix any identified issues. 
image

### CODE REVIEW

File: CONTRIBUTING.md
Comment (magnologan) [inline CONTRIBUTING.md:3]: ..aspire..

File: admin_bill_of_rights.md
Comment (magnologan) [inline admin_bill_of_rights.md:15]: Are you sure this is correct? 🤔

File: admin_bill_of_rights.md
Comment (magnologan) [inline admin_bill_of_rights.md:21]: cloud

File: CONTRIBUTING.md
Comment (TheFoxAtWork) [inline CONTRIBUTING.md:3]:
```suggestion
We aspire to create a welcoming environment for collaboration on this project
```

**Review by TheFoxAtWork — COMMENTED** [review]
I'm not sure this is good to merge in yet.  I have outstanding questions on an existing file in the repo.

File: CONTRIBUTING.md
Comment (JohnHillegass) [reply CONTRIBUTING.md]: Ah yes, I had changed this just to show how the CI would pick up the typo in https://github.com/JohnHillegass/sig-security/runs/1601873376?check_suite_focus=true

Comment (JohnHillegass) [reply CONTRIBUTING.md]: Ah yes, I had changed this just to show how the CI would pick up the typo in https://github.com/JohnHillegass/sig-security/runs/1601873376?check_suite_focus=true. Resolved by @TheFoxAtWork in 51ec618.

File: admin_bill_of_rights.md
Comment (JohnHillegass) [reply admin_bill_of_rights.md]: Hey @magnologan this was another intentional typo from the testing in my fork. Thanks for catching! I have reverted for correctness.

File: admin_bill_of_rights.md
Comment (JohnHillegass) [reply admin_bill_of_rights.md]: Hey @magnologan this was another intentional typo from the testing in my fork. Thanks for catching! I have reverted for correctness.

**Comment by JohnHillegass** [comment]
Hey @TheFoxAtWork thanks for the review! Which file are you referring to?

**Comment by TheFoxAtWork** [comment]
> Hey @TheFoxAtWork thanks for the review! Which file are you referring to?

The admin-bill-of-rights.md. I'll go ahead and approve this with all the fixes.  I have a separate discussion with the co-chairs and TLs on the fil origin.

**Review by TheFoxAtWork — APPROVED** [review]
Lgtm

**Comment by TheFoxAtWork** [comment]
I'd like another TL or Co-chair to approve this as this is a change/update to our process

**Review by JonZeolla — APPROVED** [review]

**Comment by TheFoxAtWork** [comment]
@lumjjb @ultrasaurus @JustinCappos @pragashj Need a second codeowner to take a look at this since this is a change to our existing process and will need some communication/docs in the repo on how it works.

**Comment by lumjjb** [comment]
I am in favor of adding the implementation in now but have it be passive first (not activating any gates), this will not change the current workflow and will allow us to monitor its impact on new PRs, etc. And when we think the checks are consistent with what we want for the contribution guidelines then we can add the doc and flip the switch.

**Review by lumjjb — APPROVED** [review]
LGTM! This is great work!

File: link-check-config.json
Comment (lumjjb) [inline link-check-config.json:1]: Can we move this file into `ci/` as well?

Comment (JohnHillegass) [reply link-check-config.json]: Thanks for the review @lumjjb. Moved in affb4df65436c5a6acd3a02cd0bac4ae8e1dd865.

**Comment by JohnHillegass** [comment]
@lumjjb @TheFoxAtWork anything else needed before merging this in to start the new checks??

**Comment by lumjjb** [comment]
@JohnHillegass @JonZeolla This looks great - sorry this PR seemed to have gotten lost at some point. I am still in favor of disabling the action first - if you can perhaps just comment out the file. It is likely that a lot in the repo needs to be done in order to satisfy these checks, so at least we can do the corrections offline first and get the repo in a good state before enabling the action.

Would either of you be able to do that quick change? Thanks

cc: @ultrasaurus

**Comment by JohnHillegass** [comment]
Hey @lumjjb sure I could comment out but might be easier to just disable it on the actions page, kind of like

image

that way its easy to flip on and off as needed

**Comment by lumjjb** [comment]
That looks good, all set then, i'll disable it from the UI

File: ci/setup.sh
Comment (php-coder) [inline ci/setup.sh:6]: We wouldn't need to create a temporary file if we reorder the commends:
```diff
-git diff --name-only master $HEAD > modified
 echo "these files have changed"
+git diff --name-only master $HEAD
-cat modified
```

File: ci/spelling.sh
Comment (php-coder) [inline ci/spelling.sh:5]: We don't need to execute `npm install`  when we use `npx` as the latter installs dependencies automatically. Quote from https://www.npmjs.com/package/npx

> Executes <command> either from a local node_modules/.bin, or from a central cache, installing any packages needed in order for <command> to run

File: ci/setup.sh
Comment (JohnHillegass) [reply ci/setup.sh:6]: Ah yes, good point. Thanks for pointing out. We could even possible log the file names by piping to echo.

File: ci/spelling.sh
Comment (JohnHillegass) [reply ci/spelling.sh:5]: Hmm agreed that this script may be more explicit than it needs to be. Thanks for the review!

**Pull request merged**
